lib/internal/asyncEachOfLimit.js

```diff
@@ -0,0 +1,59 @@
+const breakLoop = require('./breakLoop')
+
+// for async generators
+export default function asyncEachOfLimit(generator, limit, iteratee, callback) {
+    let done = false
+    let canceled = false
+    let awaiting = false
+    let running = 0
+    let idx = 0
+
+    function replenish() {
+        //console.log('replenish')
+        if (running >= limit || awaiting) return
```

Comment:
How do you end up with more than one generator running asynchronously with this code? If I understand the code correctly, `replenish` only gets called

1) to spawn the generator when `asyncEachOfLimit` is called
2) whenever a generator completes
3) after `iterateeCallback` is called

Do we not need to loop `limit - running` inside of replenish to call several generators at once?

Reply:
You'd only end up with more than one iteratee running if your iteratee takes longer to process than your generator takes to generate.   For example, in the tests, I have the generator ticking every 1 ms, but have the iteratee take 5 ms.

Async generators are like streams -- you get a linear sequence of items.  You can't await multiple items at the same time (and if you did, you'd get the same item multiple times, as you do when you call `.then()` on the same promise multiple times).  This is why we can't loop in `replenish` -- we'd end up calling the iteratee with the same item multiple times.

Reply:
I see thanks for explaining, I haven't used generators before.

So for my understanding, the replenish inside of the generator's `then` function is to spawn the asynchronous processes up until `limit` are running and the `replenish` inside of the `iterateeCallback` is to trigger additional items to run if the limit was reached?

Reply:
> You can't await multiple items at the same time (and if you did, you'd get the same item multiple times...)

Are you sure of this, as is right now we have the possibility of calling `next` multiple times before the promise resolves if `iteratee` is synchronous. In such a case will things break?

Reply:
> This is why we can't loop in replenish -- we'd end up calling the iteratee with the same item multiple times.

Sorry, I may be misunderstanding something, but I thought calls to `next` are queued and resolved separately. I'm looking at the `Async iterators and async iterables` section of the [proposal](https://github.com/tc39/proposal-async-iteration#async-iterators-and-async-iterables).

That being said, I think we should still probably avoid looping in replenish as we still have to wait for earlier promises to resolve.

Reply:
Ah, you're right, I was confusing `.next()` with `.then()`.  We can call `.next()` as many times as we need to, perhaps up to the concurrency limit.